### PR TITLE
[Quant] Add fused linear-leaky_relu op for onednn backend

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89188
* #88923
* #88879
* #88668
* #88665
* #88661
* __->__ #88478

**Summary**
Post op fusion can reduce data movement overhead and improve inference performance. This PR adds fused `linear-leaky_relu` op for `onednn` backend, which will be used for int8 inference with `onednn` backend. Cannot call this op with other quantization backends otherwise an error is thrown.

**Test Plan**
python test_quantization.py TestQuantizedLinear

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @leslie-fang-intel @VitalyFedyunin @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by Xia-Weiwen** [comment]
This is part of replacement of https://github.com/pytorch/pytorch/pull/76424 which is too big to land.

**Review by jgong5 — APPROVED** [review]

File: test/quantization/core/test_quantized_op.py [hunk 1]
```diff
@@ -3663,14 +3656,34 @@ def test_qlinear(self, batch_size, input_channels, output_channels, use_bias,
...
+                Y_fp32_ref = F.leaky_relu(Y_fp32_ref, **post_op_kwargs)
             Y_q_ref2 = torch.quantize_per_tensor(
                 Y_fp32_ref, Y_scale, Y_zp, dtype)
             # Assert equal
             np.testing.assert_array_almost_equal(
                 Y_q_ref2.int_repr().numpy(), Y_q.int_repr().numpy(), decimal=decimal_val)
 
+    """Tests the correctness of the quantized linear and linear_relu op."""
+    @override_qengines
+    def test_qlinear(self):
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:3671]: looks like this is qlinear + qlinear relu? I feel we can split this to two tests? we would like to make each test a unit test I think

**Review by jerryzh168 — APPROVED** [review]
lg, please split the test

**Comment by jerryzh168** [comment]
Also for future PRs, please add a "Summary" that talks about the motivation and context for the change and a "Test Plan" section that talks about how to run the test, e.g.: https://github.com/pytorch/pytorch/pull/88620

File: test/quantization/core/test_quantized_op.py (hunk 1 above)
Comment (Xia-Weiwen) [reply test/quantization/core/test_quantized_op.py:3671]: Ok I will do that.

**Comment by Xia-Weiwen** [comment]
> Also for this one and future PRs, please add a "Summary" that talks about the motivation and context for the change and a "Test Plan" section that talks about how to run the test, e.g.: #88620

Sure. I will add description later.

File: test/quantization/core/test_quantized_op.py (hunk 1 above)
Comment (Xia-Weiwen) [reply test/quantization/core/test_quantized_op.py:3671]: I have split it to two separate tests.

**Comment by Xia-Weiwen** [comment]
> > Also for this one and future PRs, please add a "Summary" that talks about the motivation and context for the change and a "Test Plan" section that talks about how to run the test, e.g.: #88620
> 
> Sure. I will add description later.

I have added summary and test plan.

**Review by jerryzh168 — APPROVED** [review]
looks good to me

**Comment by jerryzh168** [comment]
btw the Summary and Test Plan should appear after the stack..

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168.  The separated `test_qlinear` and `test_qlinear_relu` failed but they look good when running locally. I also created a draft PR which only separated the two tests without other changes, just for test. But they also failed.
Do you have any ideas why?

**Comment by Xia-Weiwen** [comment]
> Hi @jerryzh168. The separated `test_qlinear` and `test_qlinear_relu` failed so merge failed. But they look good when running locally. I also created a draft PR which only separated the two tests without other changes, just for test (#89678) But they also failed. Do you have any ideas why?

Hi, @jerryzh168. It was due to some weird rounding errors, I think. `Y_scale = 125.1234` caused the issue in fbgemm. So, I changed the scale value to `12.34`. Now all checks have passed. How does that sound to you?

File: aten/src/ATen/native/quantized/library.cpp [hunk 2]
```diff
@@ -152,6 +152,7 @@ TORCH_LIBRARY(quantized, m) {
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_relu_dynamic(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack, bool reduce_range=False) -> Tensor Y"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_dynamic_fp16(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack) -> Tensor Y"));
   m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_relu_dynamic_fp16(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack) -> Tensor Y"));
+  m.def(TORCH_SELECTIVE_SCHEMA("quantized::linear_leaky_relu(Tensor X, __torch__.torch.classes.quantized.LinearPackedParamsBase W_prepack, float Y_scale_i, int Y_zero_point_i, float negative_slope) -> Tensor Y"));
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/library.cpp:155]: nit: what does `i` in `Y_scale_i` mean?

**Comment by jerryzh168** [comment]
yeah LGTM

**Comment by jerryzh168** [comment]
> 
> 
> It has been good before we separated `test_qlinear` and `test_qlinear_relu` but the issue occurred after the separation.
> To avoid the issue and keep the test cases reasonable, I changed the scale value to `12.34`. Now all checks have passed. How does that sound to you?

what is the error? is it a numerical error or something else?

File: aten/src/ATen/native/quantized/library.cpp (hunk 2 above)
Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/library.cpp:155]: > nit: what does `i` in `Y_scale_i` mean?

Good question. I don't know. I just copied the arg name from above to look aligned https://github.com/pytorch/pytorch/blob/55b10e6b1d52aeae536129d152b1d057f159bfc0/aten/src/ATen/native/quantized/library.cpp#L150
Shall I keep it or remove it?

**Comment by Xia-Weiwen** [comment]
> > It has been good before we separated `test_qlinear` and `test_qlinear_relu` but the issue occurred after the separation.
> > To avoid the issue and keep the test cases reasonable, I changed the scale value to `12.34`. Now all checks have passed. How does that sound to you?
> 
> what is the error? is it a numerical error or something else?

Yes, it's a numerical error of fbgemm.

**Pull request closed without merging**